### ISSUE TITLE
Plugin API hangs/crashes when instancing or detaching variants, causing referential integrity error on variant-name

### ISSUE THREAD

**Issue by elhombretecla** [uxbox/uxbox#10099]
## Steps To Reproduce

1. Open this file in Penpot:

   ```text
   https://design.penpot.dev/#/workspace?team-id=b62e1aa4-d9a7-8147-8005-2813bed4056e&file-id=9edb1376-b0d6-80fd-8008-27777f7c7501&page-id=dd040aa8-5c96-8078-8008-1e71fc4e56d2&layout=layers
   ```

2. From a plugin/MCP session, run code that creates an instance from a variant component and detaches it.

   Example that previously worked for `Checkbox / Disabled Unchecked`:

   ```js
   // Re-create Checkbox / Disabled Unchecked
   // step 1: detached instance, styled, positioned

   const vc = penpotUtils.findShape(
     s => s.isVariantContainer && s.isVariantContainer() && s.name === "Checkbox"
   );

   const src = vc.variants
     .variantComponents()
     .find(c => c.variantProps.State === "Unchecked");

   const inst = src.instance();
   inst.detach();

   inst.x = 1400;
   inst.y = 146;

   inst.applyToken(
     penpotUtils.findTokenByName("color.disabled-bg"),
     ["fill"]
   );

   await new Promise(r => setTimeout(r, 150));

   storage.pendingBoardId = inst.id;

   return {
     boardId: inst.id,
     tokens: inst.tokens
   };
   ```

3. Repeat the same pattern with another variant component, for example `Radio`.

4. Try the same pattern again with `Checkbox`.

5. A smaller reproduction that may isolate the hanging line:

   ```js
   const vc = penpotUtils.findShape(
     s => s.isVariantContainer && s.isVariantContainer() && s.name === "Checkbox"
   );

   const src = vc.variants
     .variantComponents()
     .find(c => c.variantProps.State === "Unchecked");

   const inst = src.instance(); // May hang here
   inst.detach();              // Or may hang here

   return inst.id;
   ```

6. As a comparison test, run the same flow with a regular non-variant component, for example `Tooltip`:

   ```js
   const c = penpot.library.local.components.find(c => c.name === "Tooltip");
   const i = c.instance();
   i.detach();
   return i.id;
   ```

---

## Expected behavior

Creating an instance from a variant component and detaching it through the plugin API should resolve normally.

The operation should either:

* successfully create and detach the instance, returning the new shape/component ID; or
* fail with a clear JavaScript/API exception that can be caught and debugged.

The plugin session should remain alive, and recent document mutations should either be persisted correctly or rolled back in a controlled way.

---

## Actual behavior

The operation sometimes hangs indefinitely when mutating components or variants.

Observed behavior:

* No JavaScript exception is thrown.
* The promise never resolves.
* The MCP call times out after around 30 seconds.
* The plugin session remains dead for roughly 40 additional seconds.
* After reconnecting, mutations from the previous seconds are reverted.
* This looks like persistence may be debounced, and the crash/error happens before the pending mutations are flushed.
* Read-only calls do not fail. Calls such as `find`, `variantComponents`, `tokenOverview`, and export operations keep working.
* The issue appears tied to mutations involving components/variants, such as:

  * `src.instance()`
  * `detach()`
  * `createComponent`
  * `setVariantProperty`
  * `main.name = ...`

The same call, character by character, worked a few minutes earlier. It created the component and returned tokens similar to:

```js
{
  fill: "color.disabled-bg",
  ...
}
```

Then the same pattern failed with `Radio`, and afterwards it also started failing with `Checkbox`. So the behavior does not appear deterministic from the plugin API side.

The frontend eventually reports a referential integrity validation error:

```clojure
{:type :validation,
 :code :referential-integrity,
 :hint "error on validating file referential integrity",
 :file-id #uuid "9edb1376-b0d6-80fd-8008-27777f7c7501",
 :details
 [{:code :variant-bad-variant-name,
   :hint "Variant 56036d2b-ae40-8025-8008-1eb4804da6be has an invalid variant-name",
   :shape
   {:id #uuid "56036d2b-ae40-8025-8008-1eb4804da6be",
    :type :frame,
    :name "Checkbox",
    :variant-name "Disabled",
    :component-root true,
    ...},
   :file-id #uuid "9edb1376-b0d6-80fd-8008-27777f7c7501",
   :page-id #uuid "dd040aa8-5c96-8078-8008-1e71fc4e56d2",
   :shape-id #uuid "56036d2b-ae40-8025-8008-1eb4804da6be"}],
 :uri "https://design.penpot.dev/api/main/methods/update-file?id=9edb1376-b0d6-80fd-8008-27777f7c7501",
 :status 400}
```

---

## Screenshots or video

Not available.

---

## Desktop

Not available.

---

## Smartphone

Not applicable.

---

## Environment

```text
Host: https://design.penpot.dev
Version: 2.16.0-RC10-4-gf7c5ce7ac9

Profile ID: 6e80f6b0-a01e-11ea-ba1b-c554cc60e361
Team ID: b62e1aa4-d9a7-8147-8005-2813bed4056e
File ID: 9edb1376-b0d6-80fd-8008-27777f7c7501
Page ID: dd040aa8-5c96-8078-8008-1e71fc4e56d2
```

---

## Frontend Stack Trace

```console
Context:
--------------------
Hint:     error on validating file referential integrity
Prof ID:  6e80f6b0-a01e-11ea-ba1b-c554cc60e361
Team ID:  b62e1aa4-d9a7-8147-8005-2813bed4056e
File ID:  9edb1376-b0d6-80fd-8008-27777f7c7501
Version:  2.16.0-RC10-4-gf7c5ce7ac9
HREF:     https://design.penpot.dev/#/workspace?team-id=b62e1aa4-d9a7-8147-8005-2813bed4056e&file-id=9edb1376-b0d6-80fd-8008-27777f7c7501&page-id=dd040aa8-5c96-8078-8008-1e71fc4e56d2&layout=layers

====================
Data:
--------------------
{:type :validation,
 :code :referential-integrity,
 :hint "error on validating file referential integrity",
 :file-id #uuid "9edb1376-b0d6-80fd-8008-27777f7c7501",
 :details
 [{:code :variant-bad-variant-name,
   :hint "Variant 56036d2b-ae40-8025-8008-1eb4804da6be has an invalid variant-name",
   :shape
   {:id #uuid "56036d2b-ae40-8025-8008-1eb4804da6be",
    :type :frame,
    :y 162.00000223440986,
    :hide-fill-on-export false,
    :layout-gap-type :multiple,
    :layout-padding {#, #, #, #},
    :transform ##object[$app$common$geom$matrix$Matrix$$]{#, #, #, #, #, #},
    :rotation 0,
    :layout-wrap-type :nowrap,
    :layout :flex,
    :hide-in-viewer true,
    :name "Checkbox",
    :layout-align-items :center,
    :variant-name "Disabled",
    :width 18.00000286102295,
    :layout-padding-type :simple,
    :touched #{},
    :points [# # # #],
    :fixed-scroll false,
    :r2 3,
    :component-root true,
    :layout-item-h-sizing :fix,
    :proportion-lock false,
    :layout-gap {#, #},
    :transform-inverse ##object[$app$common$geom$matrix$Matrix$$]{#, #, #, #, #, #},
    ...},
   :file-id #uuid "9edb1376-b0d6-80fd-8008-27777f7c7501",
   :page-id #uuid "dd040aa8-5c96-8078-8008-1e71fc4e56d2",
   :shape-id #uuid "56036d2b-ae40-8025-8008-1eb4804da6be"}],
 :uri "https://design.penpot.dev/api/main/methods/update-file?id=9edb1376-b0d6-80fd-8008-27777f7c7501",
 :status 400}

Trace:
--------------------
Error: http error
  at new $APP.$cljs$core$ExceptionInfo$$ (https://design.penpot.dev/js/shared.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:1407:120)
  at $APP.$cljs$core$ex_info$cljs$0core$0IFn$0_invoke$0arity$03$$ (https://design.penpot.dev/js/shared.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:1408:358)
  at $APP.$cljs$core$ex_info$cljs$0core$0IFn$0_invoke$0arity$02$$ (https://design.penpot.dev/js/shared.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:1408:131)
  at $APP.$app$main$repo$handle_response$$ (https://design.penpot.dev/js/shared.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:7188:410)
  at https://design.penpot.dev/js/shared.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:3696:264
  at f (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:16480)
  at e._ [as _nextOverride] (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:16430)
  at e.WAr [as _next] (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:1877)
  at e.next (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:1371)
  at e.next [as _nextOverride] (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:25768)
  at e.WAr [as _next] (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:1877)
  at e.next (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:1371)
  at e.next [as _nextOverride] (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:14021)
  at e.WAr [as _next] (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:1877)
  at e.next (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:1371)
  at e.next [as _nextOverride] (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:14021)
  at e.WAr [as _next] (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:1877)
  at e.next (https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:1371)
  at https://design.penpot.dev/js/libs.js?version=2.16.0-RC10-4-gf7c5ce7ac9-1781039720:807:4481
```

---

## Backend Stack Trace

```console
Not available.
```

---

## Additional context

The issue seems related to plugin mutations involving variant components.

Read-only plugin/API calls continue to work reliably:

```text
find
variantComponents
tokenOverview
export
```

The calls that hang are the ones that mutate components or variants:

```text
src.instance() + detach()
createComponent
setVariantProperty
main.name = ...
```

A working call with `Checkbox` created the component and returned the expected token mapping. Shortly afterwards, the same pattern failed when repeated with `Radio`, and then started failing with `Checkbox` too. This suggests the problem may not be deterministic at the API call level.

The reported referential integrity error points to:

```text
variant-bad-variant-name
variant-name: "Disabled"
shape name: "Checkbox"
```

Hypothesis: `instance()` / `detach()` on variant components may be triggering the same variant synchronization or validation path used by `setVariantProperty` and renaming operations such as `State=...`. That path may be entering an invalid intermediate state or blocking before persistence is flushed.

This hypothesis also matches the observed pattern: the previous session created around 180 regular components in this file without this kind of hang, while the failures started when operating on variant components.

A useful manual debugging path would be to open the browser console and run the minimal reproduction line by line, checking whether the hang happens at:

```js
const inst = src.instance();
```

or at:

```js
inst.detach();
```

Another useful comparison is the same instance/detach flow with a regular non-variant component, such as `Tooltip`.

**Comment by Alotor** [comment]
Executing the following script in the web console will reproduce the error:

```
/**
 * "Plugin API hangs/crashes when instancing or detaching variants,
 *  causing referential integrity error on variant-name"
 *
 * Root cause: `component.instance()` always instantiates the copy at the
 * point (0, 0). The destination frame is chosen purely by position, and on
 * buggy builds a variant container is a valid candidate. If a variant
 * container overlaps (0, 0), the new copy becomes a direct child of the
 * container — an invalid state, since variant container children must be
 * variant mains. When the batched persist fires (e.g. on the next detach),
 * the backend rejects the update with :referential-integrity /
 * :variant-bad-variant-name, the workspace crashes to the Internal Error
 * page, the plugin connection dies (the apparent "hang") and all the
 * mutations roll back.
 *
 * How to run: paste into the plugin console / execute as plugin code in an
 * empty page, with `content:write` and `library:write` permissions.
 *
 * Expected on a buggy build: the workspace crashes with a referential
 * integrity error shortly after the script finishes, and none of the
 * changes persist after reload.
 *
 * Expected on a fixed build: the script logs "DONE", the instance lands
 * outside the variant container, and everything persists.
 */

const penpot = ɵcontext;
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

function makeBoard(name, x, y, fillColor) {
  const board = penpot.createBoard();
  board.name = name;
  board.x = x;
  board.y = y;
  board.resize(120, 60);

  const rect = penpot.createRectangle();
  board.appendChild(rect);
  rect.resize(100, 40);
  rect.x = x + 10;
  rect.y = y + 10;
  rect.fills = [{ fillColor, fillOpacity: 1 }];

  return board;
}

async function repro() {
  // 1. Two boards turned into components
  const b1 = makeBoard("Button", 0, 0, "#4f9cf9");
  const b2 = makeBoard("Button", 0, 100, "#f94f4f");

  const c1 = penpot.library.local.createComponent([b1]);
  const c2 = penpot.library.local.createComponent([b2]);

  // 2. Combine them as variants -> creates a variant container wrapping both
  const container = penpot.createVariantFromComponents([
    c1.mainInstance(),
    c2.mainInstance(),
  ]);

  // combine-as-variants is processed asynchronously; wait for the container
  // shape to materialize before touching it
  await sleep(1000);

  // 3. Make sure the variant container overlaps (0, 0), where instance()
  //    drops new copies. (Usually it already does, this makes it
  //    deterministic.)
  container.x = -50;
  container.y = -50;
  await sleep(500);

  console.log(
    "variant container:",
    container.id,
    "at",
    container.x,
    container.y,
    container.width,
    container.height
  );

  // 4. Instantiate a variant component. On buggy builds the copy becomes a
  //    direct child of the variant container.
  const inst = c1.instance();
  await sleep(500);
  console.log(
    "instance parent is the variant container?",
    inst.parent && inst.parent.id === container.id,
    "(true = bug reproduced)"
  );

  // 5. Detach it. This flushes the batched persist; on buggy builds the
  //    backend rejects it (:referential-integrity / variant-bad-variant-name)
  //    and the workspace crashes.
  inst.detach();
  await sleep(2000);

  console.log("DONE — if you can read this and the workspace did not crash, the bug is fixed");
}

repro().catch((err) => console.error("repro failed:", err));
```
